### PR TITLE
[PoC] Reactive SAUL: saul_observer based on msg_bus

### PR DESCRIPTION
**Pull request by jue89**
### Contribution description

After some bikeshedding in #14121, I started implementing a PoC.

This is the first PoC demonstrating an implementation based on `msg_bus`.

Some thoughts from my point-of-view:
- (+) Easy interface - especially if the interested thread wants to monitor many SAUL devices. No special list-item must be held in memory for every subscription.
- (+) Very few lines of code: Only ~50 LoCs are required for the basic functionality.
- (-) If the message-based IPC is used for other purposes, message types (they hold the bus ID and event type) may collide with other messages unrelated to `saul_observer`. Furthermore, message types used by `saul_observer` are not predictable as they encode the bus ID.
- (-) If more than 32 SAUL devices are present, `saul_observer_unsubscribe()` leads to unexpected behavior.

I'll come up with a second PoC show-casing a callback-based approach.

### Testing procedure

Flash `tests/saul_observer` onto a board with `saul_gpio` support and play around with the registered SAUL devices. You should see some change notifications.

```
# make -C tests/saul_observer BOARD=samr30-xpro flash term
> saul
2020-05-31 16:50:28,802 #  saul
2020-05-31 16:50:28,805 # ID	Class		Name
2020-05-31 16:50:28,808 # #0	ACT_SWITCH	LED(green)
2020-05-31 16:50:28,810 # #1	ACT_SWITCH	LED(orange)
2020-05-31 16:50:28,811 # #2	SENSE_BTN	Button(SW0)
> saul write 0 1
2020-05-31 16:50:43,470 #  saul write 0 1
2020-05-31 16:50:43,473 # Writing to device #0 - LED(green)
2020-05-31 16:50:43,475 # Data:	              1 
2020-05-31 16:50:43,478 # data successfully written to device #0
> 2020-05-31 16:50:43,484 #  SAUL DEV 0 (LED(green)) CHANGED - Data:	              1 
2020-05-31 16:50:55,254 # SAUL DEV 2 (Button(SW0)) CHANGED - Data:	              1 
2020-05-31 16:50:55,402 # SAUL DEV 2 (Button(SW0)) CHANGED - Data:	              0 
```

(Of course, no de-bouncing for the switch at the current state. This should be included for a PR that is about to be merged ...)

### Issues/PRs references

#14121

### CODE REVIEW

**Comment by maribu** [comment]
This is still interesting.

@jue89: Still PoC state, or ready for review?

**Comment by jue89** [comment]
It's still a PoC. After some testing and experimenting, I came to the conclusion that `saul_observe` should be implemented using a linked-list of callbacks. The callbacks may send a `msg_t` to a thread or a bus.

But, yes I still need this, too!

File: sys/include/saul_observer.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ * @brief   Export the SAUL observer pid
+ */
+extern kernel_pid_t saul_observer_pid;
+
+/**
+ * @brief   Trigger change check of a SAUL device
+ *
+ * @param[in] dev       pointer to the device's SAUL registry entry
+ */
+void saul_observer_queue_check(saul_reg_t *dev);
```
[the remark this answers is not in the document]
Comment (jue89) [reply sys/include/saul_observer.h:72]: Good suggestion. I adapted that change in #16277.

File: sys/include/saul_observer.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @brief   Subscribe to a SAUL device observer
+ *
+ * @param[in] entry     pointer to the attached message bus entry
+ * @param[in] dev       SAUL device to observer
+ */
+static inline void saul_observer_subscribe(msg_bus_entry_t *entry, const saul_reg_t *dev)
+{
+    msg_bus_subscribe(entry, dev->id % 32);
```
[the remark this answers is not in the document]
Comment (jue89) [reply sys/include/saul_observer.h:82]: `saul_observer_msg_bus_post()` form #16277 groups the devices by category.

**Comment by jue89** [comment]
I'm closing this PR in favor of #16277.

**Pull request closed without merging**